[Quant] Remove all the dequant nodes when the ref module has multi input args

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90605
* __->__ #90157

**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. We add a check in this PR to ensure this is the case for `_lower_static_weighted_ref_module` pass.

**Test Plan**:
We only add a check in this PR, there is no new added test case.

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

leslie-fang-intel committed:
If the ref module has multi args, remove all the dequant node as the input

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@jerryzh168 I think this PR is ready for review, could you help to take a look?

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, Could you help to take a look of this fix?

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -510,9 +510,13 @@ def _lower_static_weighted_ref_module(
-        dq_node = ref_node.args[0]
-        assert(isinstance(dq_node, Node))
-        ref_node.replace_input_with(dq_node, dq_node.args[0])
+        for arg in ref_node.args:
```
I feel this might be a bit hacky, generally we want to identify a specific pattern and just lower that pattern. this function (_lower_static_weighted_ref_module) is assuming that we have "dq -> ref_fp32_module -> q" pattern I think, could you 
1). add some checks to this function to make sure this is the case, e.g. check len(ref_node.args) == 1 or something
2). add another lowering function for conv -> add that uses this code?

leslie-fang-intel replied:
Thanks for the suggestions, @jerryzh168. Followed up these 2 steps:
1. In this PR, I only add the the check of `len(ref_node.args) == 1` for this pass ( `_lower_static_weighted_ref_module`)
2. I have added another lowering pass for `conv -> add` named as `_lower_static_weighted_ref_module_with_two_dq_inputs` in this https://github.com/pytorch/pytorch/pull/91153. Although it is working in progress to cover more cases, could you take a look of this lowering pass we talked about here? Does it look good to you?

jerryzh168 replied:
sounds great, could you update the summary for this PR as well

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. However, when we enable the `conv add` fusion, there will be a extra input node from `add` node besides the original input node from `conv`. Similar as did in the `_lower_quantized_binary_op` pass https://github.com/pytorch/pytorch/blob/41c3b41b92f5019f8d5e2f2846a06b87db01ca4e/torch/ao/quantization/fx/_lower_to_native_backend.py#L766-L775, We should remove all the `dequant` nodes in the  `_lower_static_weighted_ref_module` pass.

**Test Plan**:
It's a bug fix instead of a new feature. When we enable the `conv add` fusion PR later, we will add test cases accordingly.


cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 reviewed (approved):
thanks, please update the summary as well

leslie-fang-intel commented:
> thanks, please update the summary as well

Updated the summary for this PR.

leslie-fang-intel commented:
@pytorchbot rebase

leslie-fang-intel committed:
Update on "[Quant] Remove all the dequant nodes when the ref module has multi input args"


**Summary**:
When converting a ref module into a quant module, `_lower_static_weighted_ref_module` pass assumes the `ref_node` only has 1 input node, and only remove the first `dequant` node. We add a check in this PR to ensure this is the case for `_lower_static_weighted_ref_module` pass.

**Test Plan**:
We only add a check in this PR, there is no new added test case.

cc jerryzh168 jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
